app.py

```diff
@@ -699,22 +719,57 @@ def is_valid_position(x, y):
                     dist[(x, y)] = neighbor_cost
                     prev[(x, y)] = position
                     heapq.heappush(pq, (neighbor_cost, (x, y)))
-        
+
         # Reconstruct the path
         path = []
         while position != start:
             path.append(position)
             position = prev[position]
         path.append(start)
         path.reverse()
-        
+
         if target in path:
             self.log(f"Path found: {path}", print_type=PrintType.DEBUG)
             return path
         else:
             self.log("Path not found", print_type=PrintType.DEBUG)
             return []
 
+    def dijkstra_path(self, start, destination):
+        return shortest_path, cost
+
+    def build_graph_for_node(self, products):
+        return graph
+
+    def branch_and_bound(self, graph):
+        return lowest_cost_path
+
+    def customized_algorithm(self, order_list, graph):
+        if self.debug:
+            start_time = time.time()
+
+        total_cost = 0
+        lowest_cost_path = []
+        start_position = self.starting_position
+
+        for product in order_list:
+            shortest_path, cost = graph.dijkstra_path(self, start_position, product)
+            total_cost += cost
+            lowest_cost_path += shortest_path
+            start_position = product
+
+        shortest_path_back, cost_back = graph.dijkstra_path(self, start_position, lowest_cost_path[0])
+        lowest_cost_path += shortest_path_back
+        total_cost += cost_back

```

Comment:
I think we need to sync up and talk about this. The algorithm is:
```
1) Build the distance graph using All-Pair-Shortest-Path
2) Preprocess the item order list so that items on the same shelf are grouped together.
3) Then, going from in order of the list, you take the least cost edge to that item.
```

So I think you have kind of the right idea, but basically, **dijkstra** has already been ran.
So all you'd have to do is do a for loop through the order_list, which you're doing. 
But you just have to look up the `path` in the `graph` data structure:
```
path = []

prev_node = None
prev_dir = None
for i, product_id in enumerate(order_list):
    # Save first product and go to the next one
    if i == 0:
        prev_node = product_id
        # Need to figure out how to get previous direction properly
        prev_dir = "W"
        continue

    # Choose one of the access points, and get the path there
    for access_point in ('N', 'S', 'E', 'W'):
         next_node = (prev_node, product_id, prev_dir)
         path_to_next_node = graph[next_node][access_point]["path]
         path.append(path_to_next_node)

return path
```

This isn't perfect logic, but it's the general algorithm. Iterate through the product_ids and lookup the path between each node.

For example, if the path is `[A, B, D, G]`. The graph would be something like this:
```
(A, B, 'W'): {
    'N': {'location': (20, 11), 'cost': 31, 'path': [(0, 0), (0, 0), (0, 11), (20, 11)]}
},

(B, D, 'N'): {
    'E': {'location': (2, 0), 'cost': 29, 'path': [(20, 11), (0, 11), (2, 0)]}
}
```